src/libsync/vfs/xattr/vfs_xattr.cpp

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (C) by Kevin Ottens <kevin.ottens@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "vfs_xattr.h"
+
+#include <QFile>
+
+#include "syncfileitem.h"
+#include "filesystem.h"
+#include "common/syncjournaldb.h"
+
+#include "xattrwrapper.h"
+
+namespace xattr {
+using namespace OCC::XAttrWrapper;
+}
+
+namespace OCC {
+
+VfsXAttr::VfsXAttr(QObject *parent)
+    : Vfs(parent)
+{
+}
+
+VfsXAttr::~VfsXAttr() = default;
+
+Vfs::Mode VfsXAttr::mode() const
+{
+    return XAttr;
+}
+
+QString VfsXAttr::fileSuffix() const
+{
+    return QString();
+}
+
+void VfsXAttr::startImpl(const VfsSetupParams &)
+{
+}
+
+void VfsXAttr::stop()
+{
+}
+
+void VfsXAttr::unregisterFolder()
+{
+}
+
+bool VfsXAttr::socketApiPinStateActionsShown() const
+{
+    return true;
+}
+
+bool VfsXAttr::isHydrating() const
+{
+    return false;
+}
+
+Result<void, QString> VfsXAttr::updateMetadata(const QString &filePath, time_t modtime, qint64, const QByteArray &)
+{
+    FileSystem::setModTime(filePath, modtime);
+    return {};
+}
+
+Result<void, QString> VfsXAttr::createPlaceholder(const SyncFileItem &item)
+{
+    const auto path = QString(_setupParams.filesystemPath + item._file);
+    QFile file(path);
+    if (file.exists() && file.size() > 1
+        && !FileSystem::verifyFileUnchanged(path, item._size, item._modtime)) {
+        return QStringLiteral("Cannot create a placeholder because a file with the placeholder name already exist");
+    }
+
+    if (!file.open(QFile::ReadWrite | QFile::Truncate))
+        return file.errorString();
+
+    file.write(" ");
+    file.close();
+    FileSystem::setModTime(path, item._modtime);
```

Comment:
@er-vin Maybe use `updateMetadata `from the same class?

Reply:
Since it sounded like an excellent suggestion I was about to do it... as I was typing the change I realized that we probably shouldn't as it could introduce subtle bugs later on depending how we make the backend evolve.

One possibility is to have other xattr for the placeholder files (especially on Linux) and updateMetadata is supposed to assume that said metadata is already there which happens only once we do the `xattr::addNextcloudPlaceholderAttributes` call. So I think it's safer to leave it like this for now seeing how the duplication is really minimal.

I guess in the future we might have the situation of createPlaceholder just calling updateMetadata... but that'll also depend on how the xattr API behaves on Mac. We'll know that only when we'll have the Mac implementation...

Yes, better keep it like this for now...